File: backend/crypto.py
"""AES-256-GCM для паролів і TOTP-секретів на складі."""
from __future__ import annotations

import base64
import os
import secrets

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from .settings import CREDENTIALS_KEY


def _key() -> bytes:
    raw = (CREDENTIALS_KEY or "").strip()
    if not raw:
        raise ValueError("CREDENTIALS_KEY не заданий у .env")
    key = base64.b64decode(raw)
    if len(key) != 32:
        raise ValueError("CREDENTIALS_KEY має бути 32 байти в base64")
    return key


def encrypt(plain: str) -> str:
    iv = os.urandom(12)
    aes = AESGCM(_key())
    data = aes.encrypt(iv, (plain or "").encode("utf-8"), None)
    tag, body = data[-16:], data[:-16]
    return ".".join(
        base64.b64encode(part).decode("ascii")
        for part in (iv, tag, body)
    )


def decrypt(packed: str) -> str:
    parts = (packed or "").split(".")
    if len(parts) != 3:
        raise ValueError("Некоректний зашифрований формат")
    iv, tag, body = (base64.b64decode(p) for p in parts)
    aes = AESGCM(_key())
    return aes.decrypt(iv, body + tag, None).decode("utf-8")


def random_token(nbytes: int = 32) -> str:
    return secrets.token_urlsafe(nbytes)

File: src/lib/crypto.ts
import crypto from "crypto";
import { env } from "./env";

/**
 * Паролі від сервісів лежать у базі зашифрованими.
 * Ключ тримається в оточенні, окремо від дампів БД:
 * дамп без CREDENTIALS_KEY марний.
 */
function key(): Buffer {
  const k = Buffer.from(env.credentialsKey(), "base64");
  if (k.length !== 32) throw new Error("CREDENTIALS_KEY має бути 32 байти в base64");
  return k;
}

export function encrypt(plain: string): string {
  const iv = crypto.randomBytes(12);
  const c = crypto.createCipheriv("aes-256-gcm", key(), iv);
  const enc = Buffer.concat([c.update(plain, "utf8"), c.final()]);
  return [iv.toString("base64"), c.getAuthTag().toString("base64"), enc.toString("base64")].join(".");
}

export function decrypt(packed: string): string {
  const [iv, tag, data] = packed.split(".");
  const d = crypto.createDecipheriv("aes-256-gcm", key(), Buffer.from(iv, "base64"));
  d.setAuthTag(Buffer.from(tag, "base64"));
  return Buffer.concat([d.update(Buffer.from(data, "base64")), d.final()]).toString("utf8");
}

export function randomToken(bytes = 32): string {
  return crypto.randomBytes(bytes).toString("base64url");
}

/** Короткий код для привʼязки Telegram: без схожих символів */
export function shortCode(len = 6): string {
  const abc = "ACDEFGHJKLMNPQRTUVWXY34679";
  return Array.from(crypto.randomBytes(len)).map((b) => abc[b % abc.length]).join("");
}
